src/libsyntax/ast.rs

```diff
@@ -949,7 +949,7 @@ impl Expr {
             ExprKind::Array(exprs) if exprs.len() == 1 =>
                 exprs[0].to_ty().map(TyKind::Slice)?,
             ExprKind::Tup(exprs) => {
-                let mut tys = Vec::new();
+                let mut tys = Vec::with_capacity(exprs.len());
                 for expr in exprs {
                     tys.push(expr.to_ty()?);
```

Comment:
You can do this sort of thing by using  `.collect::<Result<Vec<_>, _>>()` (for `Result` element types).